flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/filemerging/FileMergingSnapshotManagerBase.java

```diff
@@ -345,6 +371,79 @@ protected abstract PhysicalFile getOrCreatePhysicalFileForCheckpoint(
     protected abstract void returnPhysicalFileForNextReuse(
             SubtaskKey subtaskKey, long checkpointId, PhysicalFile physicalFile) throws IOException;
 
+    /**
+     * The callback which will be triggered when all subtasks discarded (aborted or subsumed).
+     *
+     * @param checkpointId the discarded checkpoint id.
+     * @throws IOException if anything goes wrong with file system.
+     */
+    protected abstract void discardCheckpoint(long checkpointId) throws IOException;
+
+    // ------------------------------------------------------------------------
+    //  Checkpoint Listener
+    // ------------------------------------------------------------------------
+
+    @Override
+    public void notifyCheckpointComplete(SubtaskKey subtaskKey, long checkpointId)
+            throws Exception {
+        // does nothing
+    }
+
+    @Override
+    public void notifyCheckpointAborted(SubtaskKey subtaskKey, long checkpointId) throws Exception {
+        synchronized (lock) {
+            Set<LogicalFile> logicalFilesForCurrentCp = uploadedStates.get(checkpointId);
+            if (logicalFilesForCurrentCp == null) {
+                return;
+            }
+            Iterator<LogicalFile> logicalFileIterator = logicalFilesForCurrentCp.iterator();
+            while (logicalFileIterator.hasNext()) {
+                LogicalFile logicalFile = logicalFileIterator.next();
+                if (logicalFile.getSubtaskKey().equals(subtaskKey)
+                        && logicalFile.getLastUsedCheckpointID() <= checkpointId) {
+                    logicalFile.discardWithCheckpointId(checkpointId);
+                    logicalFileIterator.remove();
+                }
+            }
+
+            if (logicalFilesForCurrentCp.isEmpty()) {
+                uploadedStates.remove(checkpointId);
+                discardCheckpoint(checkpointId);
+            }
```

Comment:
nit: There is some overlap with the code of `notifyCheckpointSubsumed`, maybe it can be extracted into a function.